quinn-proto/src/connection/paths.rs

```diff
@@ -163,3 +177,605 @@ impl RttEstimator {
+    }
+
+    fn with_state(current_mtu: u16, state: Option<EnabledMtuDiscovery>) -> Self {
+        MtuDiscovery {
+            current_mtu,
+            state,
+            black_hole_counter: 0,
+            largest_acked_mtu_sized_packet: None,
+        }
+    }
+
+    /// Returns the amount of bytes that should be sent as an MTU probe, if any
+    pub fn poll_transmit(&mut self, now: Instant, next_packet_number: u64) -> Option<u16> {
+        self.state
+            .as_mut()
+            .and_then(|state| state.poll_transmit(now, self.current_mtu, next_packet_number))
+    }
+
+    /// Notifies the [`MtuDiscovery`] that the peer's `max_udp_payload_size` transport parameter has
+    /// been received
+    pub fn on_peer_max_udp_payload_size_received(&mut self, peer_max_udp_payload_size: u16) {
+        self.current_mtu = self.current_mtu.min(peer_max_udp_payload_size);
+
+        if let Some(state) = self.state.as_mut() {
+            // MTUD is only active after the connection has been fully established, so it is
+            // guaranteed we will receive the peer's transport parameters before we start probing
+            debug_assert!(matches!(state.phase, Phase::Initial));
+            state.peer_max_udp_payload_size = peer_max_udp_payload_size;
+        }
+    }
+
+    /// Notifies the [`MtuDiscovery`] that a packet has been ACKed
+    pub fn on_acked(&mut self, space: SpaceId, packet_number: u64, packet_bytes: u16) {
+        // MTU probes are only sent in application data space
+        if space != SpaceId::Data {
+            return;
+        }
+
+        // Update the state of the MTU search
+        if let Some(new_mtu) = self
+            .state
+            .as_mut()
+            .and_then(|state| state.on_probe_acked(packet_number))
+        {
+            self.current_mtu = new_mtu;
+            trace!(max_udp_payload_size = self.current_mtu, "new MTU detected");
+
+            // We know for sure the path supports the current MTU
+            self.black_hole_counter = 0;
+        }
+
+        // Reset the black hole counter if a packet the size of the current MTU or larger
+        // has been acknowledged
+        if packet_bytes >= self.current_mtu
+            && self
+                .largest_acked_mtu_sized_packet
+                .map_or(true, |pn| packet_number > pn)
+        {
+            self.black_hole_counter = 0;
+            self.largest_acked_mtu_sized_packet = Some(packet_number);
+        }
+    }
+
+    /// Returns the packet number of the in-flight MTU probe, if any
+    pub fn in_flight_mtu_probe(&self) -> Option<u64> {
+        match &self.state {
+            Some(EnabledMtuDiscovery {
+                phase: Phase::Searching(search_state),
+                ..
+            }) => search_state.in_flight_probe,
+            _ => None,
+        }
+    }
+
+    /// Notifies the [`MtuDiscovery`] that the in-flight MTU probe was lost
+    pub fn on_probe_packet_lost(&mut self) {
```

Comment:
Naming: maybe lose the `packet` word in both `on_probe_packet_lost()` and `on_non_probe_packet_lost()`?